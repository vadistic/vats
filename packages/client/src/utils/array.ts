export const removeElOnIndex = <T>(arr: T[], index: number) => [
  ...arr.slice(0, index),
  ...arr.slice(index + 1),
]

export const updateElOnIndex = <T>(arr: T[], index: number, value: T) => [
  ...arr.slice(0, index),
  value,
  ...arr.slice(index + 1),
]

export const transformElOnIndex = <T>(arr: T[], index: number, transform: (prev: T) => T) => [
  ...arr.slice(0, index),
  transform(arr[index]),
  ...arr.slice(index + 1),
]

export const insertElOnIndex = <T>(arr: T[], index: number, value: T) => [
  ...arr.slice(0, index),
  value,
  ...arr.slice(index),
]
