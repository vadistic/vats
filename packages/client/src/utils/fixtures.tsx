import { css } from '@emotion/core'
import { Fabric } from 'office-ui-fabric-react'
import React, { Suspense } from 'react'
import { ApolloProvider } from 'react-apollo'
import { ApolloProvider as ApolloHooksProvider } from 'react-apollo-hooks'
import { client } from '../apollo'
import { Layout, LoadingSpinner } from '../components'
import { theme, ThemeProvider } from '../styles'

export const random = (max: number) => Math.round(Math.random() * max)

export const AppWrapper: React.FC = ({ children }) => {
  return (
    <>
      <ApolloProvider client={client}>
        <ApolloHooksProvider client={client}>
          <ThemeProvider theme={theme}>
            <Fabric>{children}</Fabric>
          </ThemeProvider>
        </ApolloHooksProvider>
      </ApolloProvider>
    </>
  )
}

const componentWrapperStyles = css`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  width: 100%;

  & > div {
    max-width: 100%;
  }

  /* for centering */
  min-height: 80vh;
`

export const ViewWrapper: React.FC = ({ children }) => <Layout>{children}</Layout>

export const ComponentWrapper: React.FC = ({ children }) => (
  <div css={componentWrapperStyles}>
    <Suspense fallback={<LoadingSpinner label="Fetching stuff..." />}>{children}</Suspense>
  </div>
)

export const TestWrapper: React.FC = ({ children }) => (
  <AppWrapper>
    <Suspense fallback={<p>fallback</p>}>{children}</Suspense>
  </AppWrapper>
)
