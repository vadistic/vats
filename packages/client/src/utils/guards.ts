import { DeepAnyTypename } from './types'

export const isArrEmpty = (arr: Array<unknown>): arr is [] => arr.length === 0

export const isObjEmpty = (obj: object): obj is {} => Object.keys(obj).length === 0

export const anyTypename = <T>(data: T) => (data as unknown) as DeepAnyTypename<T>

export const filterNull = <T>(arr: T[]) =>
  arr.filter(el => el && el !== null) as Array<NonNullable<T>>

export const hasKey = <O>(obj: O, key: string | number | symbol): key is keyof O => key in obj

// https://github.com/Microsoft/TypeScript/issues/29780
export const tuplify = <T extends [any] | any[]>(tuple: T): T => tuple

export const XOR = (a: boolean, b: boolean) => (a && !b) || (!a && b)
export const XNOR = (a: boolean, b: boolean) => (a && b) || !(a || b)

/**
 * for index stringatures in object switch
 */

export type ObjectSwitch<T> = { [K in keyof T]: T[keyof T] }

export const objSwitch = <T extends {}>(cases: T) => cases as ObjectSwitch<T>
