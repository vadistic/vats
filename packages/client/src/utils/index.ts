export * from './router-config'
export * from './types'
export * from './fixtures'
export * from './path'
export * from './guards'
