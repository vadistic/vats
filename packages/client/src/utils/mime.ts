import { StringMap } from '@martin_hotell/rex-tils'

// TODO: autogenerate it?
export const mimeTypes: StringMap<string> = {
  'application/pdf': 'pdf',
  'application/json': 'json',
}
