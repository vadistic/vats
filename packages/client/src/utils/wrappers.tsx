import { css } from '@emotion/core'
import { ThemeProvider } from 'emotion-theming'
import { Fabric, getTheme } from 'office-ui-fabric-react'
import React, { Suspense } from 'react'
import { ApolloProvider } from 'react-apollo'
import { ApolloProvider as ApolloHooksProvider } from 'react-apollo-hooks'
import { client } from '../apollo'

export const AppWrapper: React.FC = ({ children }) => {
  return (
    <>
      <ApolloProvider client={client}>
        <ApolloHooksProvider client={client}>
          <ThemeProvider theme={getTheme()}>
            <Fabric>{children}</Fabric>
          </ThemeProvider>
        </ApolloHooksProvider>
      </ApolloProvider>
    </>
  )
}

const componentWrapperStyles = css`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  width: 100%;
  min-height: 80vh;
`

export const ComponentWrapper: React.FC = ({ children }) => (
  <div css={componentWrapperStyles}>
    <Suspense fallback={<span>Loading...</span>}>{children}</Suspense>
  </div>
)

export const TestWrapper: React.FC = ({ children }) => (
  <AppWrapper>
    <Suspense fallback={<span>Loading...</span>}>{children}</Suspense>
  </AppWrapper>
)
