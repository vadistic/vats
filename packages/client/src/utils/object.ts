import { XNOR, XOR } from './guards'
import { Head, Tail } from './types'

/*
 * util for safely traversing object trees
 */
export type TraversePath<State extends any, Paths extends any[]> = Head<Paths> extends keyof State
  ? {
      0: State[Head<Paths>]
      1: TraversePath<State[Head<Paths>], Tail<Paths>>
    }[Tail<Paths> extends never ? 0 : 1]
  : void

/**
 * access nested object property
 */
export const getIn = <State extends any, Paths extends Array<string | number>>(
  state: State,
  ...paths: Paths
): TraversePath<State, Paths> => {
  const [head, ...tail] = paths

  if (!state.hasOwnProperty(head)) {
    console.error(state, `has not ${head}`)
    throw new Error()
  }

  if (tail.length) {
    return getIn(state[head], ...tail)
  }

  return state[head]
}

export const tryGetIn = <State extends any, Paths extends Array<string | number>>(
  state: State,
  ...paths: Paths
): TraversePath<State, Paths> | undefined => {
  const [head, ...tail] = paths

  if (!state.hasOwnProperty(head)) {
    return undefined
  }

  if (tail.length) {
    return getIn(state[head], ...tail)
  }

  return state[head]
}

/**
 * set nested object property to value
 */

export const mutableSetValueIn = <
  State extends any,
  Value extends any,
  Paths extends Array<string | number>
>(
  state: State,
  value: Value,
  path: Paths,
) => {
  const [head, ...tail] = path

  // handle obj <=> array conversion
  const isHeadArray = Array.isArray(state[head])
  const shouldHeadBeArray = typeof tail[0] === 'number'

  if (tail.length > 0) {
    const nextHead = XNOR(shouldHeadBeArray, isHeadArray)
      ? state[head]
      : shouldHeadBeArray
      ? []
      : {}

    state[head] = mutableSetValueIn(nextHead, value, tail)
  }

  if (tail.length === 0) {
    state[head] = value
  }

  return state
}

/**
 * this one's immutable
 */
export const setValueIn = <
  State extends any,
  Value extends any,
  Paths extends Array<string | number>
>(
  state: State,
  value: Value,
  path: Paths,
) => {
  const [head, ...tail] = path

  let temp: any

  if (tail.length > 0) {
    temp = mutableSetValueIn(state[head] || typeof head === 'number' ? [] : {}, value, tail)
  }

  if (tail.length === 0) {
    temp = value
  }

  return { ...state, [head]: temp }
}
