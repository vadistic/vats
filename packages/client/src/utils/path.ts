import { getIn } from './object'
import { Head, Tail } from './types'

/*
 * util for safely building nested properties paths (for formik)
 */
export type CheckPath<State extends any, Paths extends any[]> = Head<Paths> extends keyof State
  ? {
      0: string
      1: CheckPath<State[Head<Paths>], Tail<Paths>>
    }[Tail<Paths> extends never ? 0 : 1]
  : void

type Leaf = string | number | boolean | undefined

export type CheckLeafPath<State extends any, Paths extends any[]> = Head<Paths> extends keyof State
  ? {
      0: State[Head<Paths>] extends Leaf ? string : never
      1: CheckPath<State[Head<Paths>], Tail<Paths>>
    }[Tail<Paths> extends never ? 0 : 1]
  : void

/**
 * validate type path and transform to lodash.style
 */
export const buildLoPath = (paths: Array<string | number>) => {
  const loPath = paths.reduce((acc, el, i) => {
    acc += typeof el === 'number' ? `[${el}]` : i === 0 ? el : `.${el}`

    return acc
  }, '')

  return loPath
}

export const getLoPath = <S, P extends string[]>(state: S, ...paths: P) =>
  (buildLoPath(paths) as unknown) as CheckPath<S, P>

/**
 * validate type path for leaf node and transform to lodash.style
 */
export const getLoLeafPath = <S, P extends Array<string | number>>(state: S, ...paths: P) =>
  (buildLoPath(paths) as unknown) as CheckLeafPath<S, P>

/**
 *  and now reversing lodash style paths :/
 */
export const getInByPath = (state: any, path: string) =>
  getIn(state, ...path.split(/\.|(?:\[|\])/).filter(el => !!el))

/*
 * Different approach
 * https://github.com/jaredpalmer/formik/issues/1334
 */

export interface IFieldPath {
  PATH: string
}

export type PathProxy<S> = IFieldPath & { [K in keyof S]: PathProxyValue<S[K]> }

type PathProxyValue<S> = S extends object ? PathProxy<S> : IFieldPath

const idPath = { PATH: '' } as IFieldPath

export const pathProxy = <S>(parent: IFieldPath = idPath as any): PathProxy<S> => {
  return new Proxy(parent as any, {
    get(target: any, key: any) {
      if (key in target) {
        return target[key]
      }
      return pathProxy<any>({
        PATH: `${parent.PATH ? parent.PATH + '.' : idPath.PATH}${key}`,
      } as any)
    },
  })
}
