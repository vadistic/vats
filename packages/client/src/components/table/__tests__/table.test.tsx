import React from 'react'
import { render } from 'react-testing-library'
import { TestWrapper } from '../../../utils/wrappers'
import { TableBasicFixture } from '../__jsxfixtures__/basic.jsxfixture'

it('renders without crashing', () => {
  const { asFragment } = render(
    <TestWrapper>
      <TableBasicFixture />
    </TestWrapper>,
  )

  expect(asFragment()).toMatchSnapshot()
})
