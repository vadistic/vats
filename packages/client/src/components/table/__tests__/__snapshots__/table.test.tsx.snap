// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="ms-Fabric root-47"
  >
    <div
      class="css-s2v4g3-tableStyles"
    >
      <div
        class="ms-Viewport"
        style="min-width: 1px; min-height: 1px;"
      />
    </div>
  </div>
</DocumentFragment>
`;
